File: Python/9-DynamicProgramming(DP)/Problems/3-CoinChange/CoinChange.py
def coin_change_recursive(coins, amount):
    """
    Recursive solution for the Coin Change problem.

    This is the most basic approach, but it's also the least efficient due to
    repeated calculations of the same subproblems.  It often leads to a stack
    overflow for larger amounts or a large number of coins.

    Args:
        coins: A list of coin denominations.
        amount: The target amount to make change for.

    Returns:
        The minimum number of coins needed to make up the amount, or -1 if it's
        not possible.
    """
    if amount == 0:
        return 0
    if amount < 0:
        return -1

    min_coins = float('inf')  # Initialize with a large value
    for coin in coins:
        res = coin_change_recursive(coins, amount - coin)
        if res != -1:
            min_coins = min(min_coins, 1 + res)  # 1 + res:  1 coin used + result of subproblem

    return min_coins if min_coins != float('inf') else -1

def coin_change_memoization(coins, amount):
    """
    Memoization (Top-Down Dynamic Programming) solution for the Coin Change problem.

    This approach improves upon the recursive solution by storing the results of
    solved subproblems in a memo (dictionary).  Before computing the solution to
    a subproblem, it checks if the solution is already stored.  This avoids
    redundant calculations.

    Args:
        coins: A list of coin denominations.
        amount: The target amount to make change for.

    Returns:
        The minimum number of coins needed to make up the amount, or -1 if it's
        not possible.
    """
    memo = {}  # Use a dictionary to store results of subproblems

    def _coin_change_memo(amount):
        if amount == 0:
            return 0
        if amount < 0:
            return -1
        if amount in memo:
            return memo[amount]  # Return the stored result if it exists

        min_coins = float('inf')
        for coin in coins:
            res = _coin_change_memo(amount - coin)
            if res != -1:
                min_coins = min(min_coins, 1 + res)
        memo[amount] = min_coins if min_coins != float('inf') else -1  # Store the result
        return memo[amount]

    return _coin_change_memo(amount)


def coin_change_tabulation(coins, amount):
    """
    Tabulation (Bottom-Up Dynamic Programming) solution for the Coin Change problem.

    This approach builds a table (dp array) of solutions to subproblems, starting
    from the base case (amount 0) and iteratively computing solutions for larger
    amounts.  It avoids recursion and directly accesses previously computed
    values in the table.

    Args:
        coins: A list of coin denominations.
        amount: The target amount to make change for.

    Returns:
        The minimum number of coins needed to make up the amount, or -1 if it's
        not possible.
    """
    dp = [float('inf')] * (amount + 1)  # dp[i] stores min coins for amount i
    dp[0] = 0  # Base case: 0 coins needed for amount 0

    for i in range(1, amount + 1):  # Iterate through all amounts from 1 to amount
        for coin in coins:
            if i - coin >= 0:
                dp[i] = min(dp[i], 1 + dp[i - coin])  # Transition function

    return dp[amount] if dp[amount] != float('inf') else -1

def coin_change_optimized(coins, amount):
    """
    Optimized Tabulation (Bottom-Up DP) for Coin Change.

    This is a space-optimized version of the tabulation method.  In the standard
    tabulation method, we create a DP table of size (amount + 1).  However, we
    only need the previous subproblem results to compute the current result.
    This optimized version still uses tabulation, but it does so in a way that
    reduces the space complexity.  However, for the coin change problem, the
    standard tabulation method is already quite efficient, and further
    optimization might not be necessary for most practical cases.
    Args:
        coins: List of coin denominations.
        amount: Target amount.

    Returns:
        Minimum number of coins, or -1 if not possible.
    """
    dp = [float('inf')] * (amount + 1)
    dp[0] = 0

    for i in range(1, amount + 1):
        for coin in coins:
            if i - coin >= 0:
                dp[i] = min(dp[i], 1 + dp[i - coin])
    return dp[amount] if dp[amount] != float('inf') else -1

def coin_change_iterative_dp(coins, amount):
    """
    Iterative Dynamic Programming solution for Coin Change.

    This is another iterative approach, similar to tabulation, but written slightly differently.
    It aims to be very clear and explicit in its iteration.  Like tabulation, it builds
    a table of solutions to subproblems.

    Args:
        coins: A list of coin denominations.
        amount: The target amount.

    Returns:
        The minimum number of coins needed, or -1 if not possible.
    """
    dp = [float('inf')] * (amount + 1)
    dp[0] = 0

    for current_amount in range(1, amount + 1):
        for coin in coins:
            if current_amount - coin >= 0:
                remaining_amount = current_amount - coin
                dp[current_amount] = min(dp[current_amount], 1 + dp[remaining_amount])

    return dp[amount] if dp[amount] != float('inf') else -1
# Example usage:
coins = [1, 2, 5]
amount = 11

print("Recursive:", coin_change_recursive(coins, amount))
print("Memoization:", coin_change_memoization(coins, amount))
print("Tabulation:", coin_change_tabulation(coins, amount))
print("Optimized Tabulation:", coin_change_optimized(coins, amount))
print("Iterative DP:", coin_change_iterative_dp(coins, amount))

File: C++/9-DynamicProgramming(DP)/Problems/3-CoinChange/CoinChange.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <climits>
#include <numeric> // For std::accumulate

using namespace std;

// 1. Recursive Approach (Plain Recursion)
//    - This is the most basic approach, but it's highly inefficient due to redundant calculations.
//    - It directly implements the recursive definition of the coin change problem.
int coinChangeRecursive(vector<int>& coins, int amount) {
    // Base cases:
    if (amount == 0) {
        return 0; // If the target amount is 0, no coins are needed.
    }
    if (amount < 0) {
        return -1; // If the target amount is negative, it's impossible to make change.
    }

    int minCoins = INT_MAX; // Initialize with a large value to find the minimum.

    // Try using each coin as the last coin in the change.
    for (int coin : coins) {
        int subResult = coinChangeRecursive(coins, amount - coin); // Recursive call
        if (subResult != -1) { // If a solution exists for the subproblem...
            minCoins = min(minCoins, subResult + 1); // Update minCoins
        }
    }

    return (minCoins == INT_MAX) ? -1 : minCoins; // If no solution was found, return -1.
}

// 2. Memoization (Top-Down Dynamic Programming)
//    - This approach improves upon the recursive solution by storing the results of
//      subproblems in a table (memo) to avoid recalculating them.
//    - It's still recursive but much more efficient.
int coinChangeMemo(vector<int>& coins, int amount, vector<int>& memo) {
    // Base cases:
    if (amount == 0) {
        return 0;
    }
    if (amount < 0) {
        return -1;
    }

    // Check if the result for this amount is already memoized.
    if (memo[amount] != -1) {
        return memo[amount];
    }

    int minCoins = INT_MAX;
    for (int coin : coins) {
        int subResult = coinChangeMemo(coins, amount - coin, memo);
        if (subResult != -1) {
            minCoins = min(minCoins, subResult + 1);
        }
    }

    // Store the result in the memo table before returning.
    memo[amount] = (minCoins == INT_MAX) ? -1 : minCoins;
    return memo[amount];
}

int coinChangeMemoization(vector<int>& coins, int amount) {
    vector<int> memo(amount + 1, -1); // Initialize memo table with -1 (indicating not calculated).
    return coinChangeMemo(coins, amount, memo);
}

// 3. Tabulation (Bottom-Up Dynamic Programming)
//    - This is an iterative approach that builds up a table of solutions to subproblems
//      from the bottom up.
//    - It's generally more efficient than memoization because it avoids recursion overhead.
int coinChangeTabulation(vector<int>& coins, int amount) {
    vector<int> dp(amount + 1, INT_MAX); // dp[i] stores the minimum coins needed for amount i.
    dp[0] = 0; // Base case: 0 coins are needed for amount 0.

    // Iterate through all possible amounts from 1 to the target amount.
    for (int i = 1; i <= amount; ++i) {
        // For each amount, try using each coin.
        for (int coin : coins) {
            if (i - coin >= 0 && dp[i - coin] != INT_MAX) {
                // If using this coin leads to a valid subproblem solution,
                // update the minimum number of coins needed for the current amount.
                dp[i] = min(dp[i], dp[i - coin] + 1);
            }
        }
    }

    return (dp[amount] == INT_MAX) ? -1 : dp[amount]; // Return the result for the target amount.
}

// 4. Tabulation with Coin Count Tracking
//    - This approach is similar to tabulation, but it also keeps track of the number
//      of times each coin is used in the optimal solution.
//    - This is useful if you need to know the actual coin combination.
int coinChangeTabulationWithCount(vector<int>& coins, int amount, vector<int>& coinCount) {
    vector<int> dp(amount + 1, INT_MAX);
    vector<int> lastCoin(amount + 1, -1); // Store the last coin used to reach amount i.
    dp[0] = 0;

    for (int i = 1; i <= amount; ++i) {
        for (int coin : coins) {
            if (i - coin >= 0 && dp[i - coin] != INT_MAX) {
                if (dp[i - coin] + 1 < dp[i]) {
                    dp[i] = dp[i - coin] + 1;
                    lastCoin[i] = coin; // Store the coin used.
                }
            }
        }
    }

    if (dp[amount] == INT_MAX) {
        return -1;
    }

    // Reconstruct the coin combination and count the occurrences of each coin.
    int currentAmount = amount;
    while (currentAmount > 0) {
        int coin = lastCoin[currentAmount];
        for (int i = 0; i < coins.size(); ++i)
        {
            if (coins[i] == coin)
            {
                coinCount[i]++;
                break;
            }
        }
        currentAmount -= coin;
    }
    return dp[amount];
}

// 5. Optimized 1D DP
//    - A space-optimized version of the tabulation method.  It uses only a 1D DP array.
int coinChangeOptimized(vector<int>& coins, int amount) {
    vector<int> dp(amount + 1, amount + 1);  // Initialize with a value > max possible coins
    dp[0] = 0;

    for (int i = 1; i <= amount; ++i) {
        for (int j = 0; j < coins.size(); ++j) {
            if (coins[j] <= i) {
                dp[i] = min(dp[i], dp[i - coins[j]] + 1);
            }
        }
    }
    return dp[amount] > amount ? -1 : dp[amount];
}

int main() {
    vector<int> coins = {1, 2, 5};
    int amount = 11;

    cout << "Coin Change Problem: Finding the minimum number of coins to make up a given amount." << endl;
    cout << "Coins: ";
    for (int coin : coins) {
        cout << coin << " ";
    }
    cout << endl;
    cout << "Target Amount: " << amount << endl << endl;

    // 1. Recursive Approach
    cout << "1. Recursive Approach: ";
    int resultRecursive = coinChangeRecursive(coins, amount);
    if (resultRecursive != -1) {
        cout << "Minimum coins needed: " << resultRecursive << endl;
    } else {
        cout << "No solution exists." << endl;
    }

    // 2. Memoization Approach
    cout << "2. Memoization Approach: ";
    int resultMemo = coinChangeMemoization(coins, amount);
    if (resultMemo != -1) {
        cout << "Minimum coins needed: " << resultMemo << endl;
    } else {
        cout << "No solution exists." << endl;
    }

    // 3. Tabulation Approach
    cout << "3. Tabulation Approach: ";
    int resultTabulation = coinChangeTabulation(coins, amount);
    if (resultTabulation != -1) {
        cout << "Minimum coins needed: " << resultTabulation << endl;
    } else {
        cout << "No solution exists." << endl;
    }

    // 4. Tabulation with Coin Count Tracking
    cout << "4. Tabulation with Coin Count: ";
    vector<int> coinCount(coins.size(), 0); // Initialize coin counts to 0.
    int resultTabulationWithCount = coinChangeTabulationWithCount(coins, amount, coinCount);
    if (resultTabulationWithCount != -1) {
        cout << "Minimum coins needed: " << resultTabulationWithCount << endl;
        cout << "Coin Count: ";
        for (int i = 0; i < coins.size(); ++i) {
            cout << coins[i] << ":" << coinCount[i] << " ";
        }
        cout << endl;
    } else {
        cout << "No solution exists." << endl;
    }

    // 5. Optimized 1D DP
    cout << "5. Optimized 1D DP Approach: ";
    int resultOptimized = coinChangeOptimized(coins, amount);
    if (resultOptimized != -1) {
        cout << "Minimum coins needed: " << resultOptimized << endl;
    } else {
        cout << "No solution exists." << endl;
    }

    return 0;
}

